.github/workflows/upload-win-installer.yml

```diff
@@ -0,0 +1,96 @@
+name: Upload Windows Installer
+
+on:
+  release:
+    types: [created, published, edited]
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Release version to build and upload (e.g. "4.2.1")'
+        required: true
+jobs:
+  build:
+    runs-on: windows-latest
+    env:
+        FETCH_BASE_URL: ${{ github.server_url }}/${{ github.repository }}
+    steps:
+    - name: Determine version
+      id: getversion
+      run: |
+        $version = "${{ inputs.version }}"
+        if ($version.Length -lt 1) {
+          $version = "${{ github.event.release.tag_name }}"
+          if ($version.Length -lt 1) {
+            Write-Host "Could not determine version!"
+            Exit 1
+          }
+        }
+        Write-Output "::set-output name=version::$version"
+    - uses: actions/checkout@v3
+    - name: Check
+      id: check
+      run: |
+        Push-Location contrib\win-installer
+        .\check.ps1 ${{steps.getversion.outputs.version}}
+        $code = $LASTEXITCODE
+        if ($code -eq 2) {
+          Write-Output "::set-output name=already-exists::true"
+          Pop-Location
+          Exit 0
+        }
+        Write-Output "UPLOAD_ASSET_NAME=$env:UPLOAD_ASSET_NAME" | Out-File -FilePath $env:GITHUB_ENV -Append
+        Pop-Location
+        Exit $code
+    - name: Set up Go
+      uses: actions/setup-go@v3
+      if: steps.check.outputs.already-exists != 'true'
+      with:
+        go-version: 1.18
+    - name: Setup Signature Tooling
+      if: steps.Check.outputs.already-exists != 'true'
+      run: |
+          dotnet tool install --global AzureSignTool --version 3.0.0
+          echo "CERT_NAME=${{secrets.AZ_CERT_NAME}}" | Out-File -FilePath $env:GITHUB_ENV -Append
+          echo "VAULT_ID=${{secrets.AZ_VAULT_ID}}" | Out-File -FilePath $env:GITHUB_ENV -Append
+          echo "APP_ID=${{secrets.AZ_APP_ID}}" | Out-File -FilePath $env:GITHUB_ENV -Append
+          echo "TENANT_ID=${{secrets.AZ_TENANT_ID}}" | Out-File -FilePath $env:GITHUB_ENV -Append
+          echo "CLIENT_SECRET=${{secrets.AZ_CLIENT_SECRET}}" | Out-File -FilePath $env:GITHUB_ENV -Append
```

Comment:
Do these certs expire? If an update/change/replacement is needed it would be good to have a reference (comment) to docs on how to do this.

Reply:
> > [NO NEW TESTS NEEDED]
> 
> Not having looked at the changes yet, the need for this magic bothers me slightly for such a large & significant feature.:
> 
> 1. It would be good to test the installer functions at PR time.  The worst possible time to find out the installer is broken is when a new release is tagged, we need to know as early as possible.  We can run [Windows VMs](https://cloud.google.com/compute/docs/images/os-details#windows_client) and [containers](https://cirrus-ci.org/guide/windows/) in CI, it's not been done to date, b/c we (in general) lack the expertise.  Maybe this is something you'd consider helping with?

I agree we should have something at PR time to verify the installer at least builds (when related installer files are touched). This is something I planned to follow-up with another change on. Unfortunately we can't (yet) test the execution of the full installer, since WSL fails without nested virt, which is not well supported on Windows with hyperscalers and cloud CI, or is cost prohibitive (e.g. GCP only supports linux with nested virt, EC2 requires you buy a high end bare metal node, GHA uses v2 non-nested nodes, etc). I have a request open with the Cirrus folks (https://github.com/cirruslabs/cirrus-ci-docs/discussions/953) to add support for direct usage of Azure VMs, which would address this problem. The same issue is also standing in the way for PR testing of podman machine on windows.  That said some basic verification is certainly doable.

Yes I am up for helping. If possible I'd like to incrementally improve on these aspects in subsequent PRs. If you prefer I could quickly turnaround and include a GHA for PR build verification temporarily. That job would then be replaced with a Cirrus task once I have the dep baseline scripted. 

> 2. If there are file additions which cannot practically be tested, please update `contrib/cirrus/pr-should-include-tests` to ignore them.  Otherwise anybody that touches them will again need the `[NO NEW TESTS NEEDED]` magic.

Will do. IIRC contrib is already excluded (will double check that), but the two added commands to cmd (only used by the installer) need to exclude the full PR run.

Reply:
Yes, there are managed/updates in a vault independently of the client secrets. However the secrets can also change as needed or in common rotation. The scripts should not need to change related to this though.  I'll ping you with more details on this.

Reply:
Okay, yes please.  My main concern is some out-of-band process breaking/failing which surprises developers by breaking their PRs.  We have a way to share secrets among the team, so maybe these details can be placed there.